File: Samples/pdf417-sample/pdf417-sample/ViewController.m
//
//  ViewController.m
//  pdf417-sample
//
//  Created by Jura on 16/07/15.
//  Copyright (c) 2015 MicroBlink. All rights reserved.
//

#import "ViewController.h"
#import "CustomOverlay.h"

#import <MicroBlink/MicroBlink.h>

@interface ViewController () <MBBarcodeOverlayViewControllerDelegate, MBCustomOverlayViewControllerDelegate, UIAlertViewDelegate>

@property (nonatomic, strong) MBBarcodeRecognizer *barcodeRecognizer;
@property (nonatomic, strong) MBPdf417Recognizer *pdf417Recognizer;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    /** First, set license key as soon as possible */
    // Valid unitl: 2018-07-02
    [[MBMicroblinkSDK sharedInstance] setLicenseResource:@"license-pdf" withExtension:@"txt" inSubdirectory:@"License" forBundle:[NSBundle mainBundle]];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


- (IBAction)didTapScan:(id)sender {
    
    /** Create recognizers */
    self.barcodeRecognizer = [[MBBarcodeRecognizer alloc] init];
    self.barcodeRecognizer.scanQR = YES;
    
    self.pdf417Recognizer = [[MBPdf417Recognizer alloc] init];
    
    MBBarcodeOverlaySettings* settings = [[MBBarcodeOverlaySettings alloc] init];
    
    NSMutableArray<MBRecognizer *> *recognizers = [[NSMutableArray alloc] init];
    
    [recognizers addObject:self.barcodeRecognizer];
    [recognizers addObject:self.pdf417Recognizer];
    
    /** Create recognizer collection */
    settings.uiSettings.recognizerCollection = [[MBRecognizerCollection alloc] initWithRecognizers:recognizers];
    
    MBBarcodeOverlayViewController *overlayVC = [[MBBarcodeOverlayViewController alloc] initWithSettings:settings andDelegate:self];
    UIViewController<MBRecognizerRunnerViewController>* recognizerRunnerViewController = [MBViewControllerFactory recognizerRunnerViewControllerWithOverlayViewController:overlayVC];
    
    /** Present the recognizer runner view controller. You can use other presentation methods as well (instead of presentViewController) */
    [self presentViewController:recognizerRunnerViewController animated:YES completion:nil];
}

- (IBAction)didTapCustomUI:(id)sender {
    /** Create recognizers */
    self.barcodeRecognizer = [[MBBarcodeRecognizer alloc] init];
    self.barcodeRecognizer.scanQR = YES;
    
    self.pdf417Recognizer = [[MBPdf417Recognizer alloc] init];
    
    MBSettings* settings = [[MBSettings alloc] init];
    
    NSMutableArray<MBRecognizer *> *recognizers = [[NSMutableArray alloc] init];
    
    [recognizers addObject:self.barcodeRecognizer];
    [recognizers addObject:self.pdf417Recognizer];
    
    /** Create recognizer collection */
    settings.uiSettings.recognizerCollection = [[MBRecognizerCollection alloc] initWithRecognizers:recognizers];
    
    CustomOverlay *overlayVC = [CustomOverlay  initFromStoryboardWithSettings:settings andDelegate:self];
    UIViewController<MBRecognizerRunnerViewController>* recognizerRunnerViewController = [MBViewControllerFactory recognizerRunnerViewControllerWithOverlayViewController:overlayVC];
    
    /** Present the recognizer runner view controller. You can use other presentation methods as well (instead of presentViewController) */
    [self presentViewController:recognizerRunnerViewController animated:YES completion:nil];
}

#pragma mark - MBBarcodeOverlayViewControllerDelegate

- (void)overlayViewControllerDidFinishScanning:(MBOverlayViewController *)overlayViewController state:(MBRecognizerResultState)state {
    /** This is done on background thread*/
    [overlayViewController.recognizerRunnerViewController pauseScanning];
    
    NSString* message;
    NSString* title;
    
    if (self.barcodeRecognizer.result.resultState == MBRecognizerResultStateValid) {
        title = @"QR code";
        
        // Save the string representation of the code
        message = [self.barcodeRecognizer.result stringData];
    }
    else if (self.pdf417Recognizer.result.resultState == MBRecognizerResultStateValid) {
        title = @"PDF417";
        
        // Save the string representation of the code
        message = [self.pdf417Recognizer.result stringData];
    }
    
    /** Needs to be called on main thread beacuse everything prior is on background thread */
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:title
                                                                                 message:message
                                                                          preferredStyle:UIAlertControllerStyleAlert];
        
        UIAlertAction* okAction = [UIAlertAction actionWithTitle:@"OK"
                                                           style:UIAlertActionStyleDefault
                                                         handler:^(UIAlertAction * _Nonnull action) {
                                                             [self dismissViewControllerAnimated:YES completion:nil];
                                                         }];
        
        [alertController addAction:okAction];
        
        [overlayViewController presentViewController:alertController animated:YES completion:nil];
    });
    
}

- (void)overlayViewControllerDidTapClose:(MBOverlayViewController *)overlayViewController {
    [self dismissViewControllerAnimated:YES completion:nil];
}
- (void)barcodeOverlayViewControllerDidFinishScanning:(nonnull MBBarcodeOverlayViewController *)barcodeOverlayViewController state:(MBRecognizerResultState)state {
    [self overlayViewControllerDidFinishScanning:barcodeOverlayViewController state:state];
}

- (void)barcodeOverlayViewControllerDidTapClose:(nonnull MBBarcodeOverlayViewController *)barcodeOverlayViewController {
    [self overlayViewControllerDidTapClose:barcodeOverlayViewController];
}

- (void)customOverlayViewControllerDidFinishScanning:(nonnull CustomOverlay *)overlayViewController state:(MBRecognizerResultState)state {
    [self overlayViewControllerDidFinishScanning:overlayViewController state:state];
}

- (void)customOverlayViewControllerDidTapClose:(nonnull CustomOverlay *)overlayViewController {
    [self overlayViewControllerDidTapClose:overlayViewController];
}

@end

File: Samples/Swift/pdf417-sample-Swift/pdf417-sample-Swift/ViewController.swift
//
//  ViewController.swift
//  pdf417-sample-Swift
//
//  Created by Dino on 17/12/15.
//  Copyright © 2015 Dino. All rights reserved.
//

import UIKit
import MicroBlink

class ViewController: UIViewController, MBCustomOverlayViewControllerDelegate, MBBarcodeOverlayViewControllerDelegate {
    
    var pdf417Recognizer : MBPdf417Recognizer?
    var barcodeRecognizer : MBBarcodeRecognizer?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Valid until: 2018-04-29
        MBMicroblinkSDK.sharedInstance().setLicenseResource("license-pdf-swift", withExtension: "txt", inSubdirectory: "License", for: Bundle.main)
    }
    
    func createScanSettings() -> MBSettings {
        /** Create barcode recognizer */
        self.barcodeRecognizer = MBBarcodeRecognizer()
        self.barcodeRecognizer?.scanPdf417 = true
        
        self.pdf417Recognizer = MBPdf417Recognizer()
        
        /** Create barcode settings */
        let settings : MBSettings = MBSettings()
        
        /** Crate recognizer collection */
        let recognizerList : Array = [self.barcodeRecognizer!, self.pdf417Recognizer!] as! [MBRecognizer]
        let recognizerCollection : MBRecognizerCollection = MBRecognizerCollection(recognizers: recognizerList)
        
        /** Add recognizer collection to barcode settings */
        settings.uiSettings.recognizerCollection = recognizerCollection
        return settings;
    }
    
    @IBAction func didTapCustomUI(_ sender: Any) {
        
        /** Create barcode recognizer */
        self.barcodeRecognizer = MBBarcodeRecognizer()
        self.barcodeRecognizer?.scanQR = true
        
        self.pdf417Recognizer = MBPdf417Recognizer()
        
        /** Create barcode settings */
        let settings : MBSettings = MBSettings()
        
        /** Crate recognizer collection */
        let recognizerList : Array = [self.barcodeRecognizer!, self.pdf417Recognizer!] as! [MBRecognizer]
        let recognizerCollection : MBRecognizerCollection = MBRecognizerCollection(recognizers: recognizerList)
        
        /** Add recognizer collection to barcode settings */
        settings.uiSettings.recognizerCollection = recognizerCollection
        
        /** Create your overlay view controller */
        let customOverlayViewController : CustomOverlay = CustomOverlay.initFromStoryboardWith(settings: settings, delegate: self)
        
        /** Create recognizer view controller with wanted overlay view controller */
        let recognizerRunneViewController : UIViewController = MBViewControllerFactory.recognizerRunnerViewController(withOverlayViewController: customOverlayViewController)
        
        /** Present the recognizer runner view controller. You can use other presentation methods as well (instead of presentViewController) */
        self.present(recognizerRunneViewController, animated: true, completion: nil)
    }
    @IBAction func didTapScan(_ sender: AnyObject) {
        
        /** Create barcode recognizer */
        self.barcodeRecognizer = MBBarcodeRecognizer()
        self.barcodeRecognizer?.scanQR = true
        
        self.pdf417Recognizer = MBPdf417Recognizer()
        
        /** Create barcode settings */
        let settings : MBBarcodeOverlaySettings = MBBarcodeOverlaySettings()
        
        /** Crate recognizer collection */
        let recognizerList : Array = [self.barcodeRecognizer!, self.pdf417Recognizer!] as! [MBRecognizer]
        let recognizerCollection : MBRecognizerCollection = MBRecognizerCollection(recognizers: recognizerList)
        
        /** Add recognizer collection to barcode settings */
        settings.uiSettings.recognizerCollection = recognizerCollection
        
        /** Create your overlay view controller */
        let barcodeOverlayViewController : MBBarcodeOverlayViewController = MBBarcodeOverlayViewController(settings: settings, andDelegate: self)
        
        /** Create recognizer view controller with wanted overlay view controller */
        let recognizerRunneViewController : UIViewController = MBViewControllerFactory.recognizerRunnerViewController(withOverlayViewController: barcodeOverlayViewController)
        
        /** Present the recognizer runner view controller. You can use other presentation methods as well (instead of presentViewController) */
        self.present(recognizerRunneViewController, animated: true, completion: nil)
    }
    
    func overlayViewControllerDidFinishScanning(_ overlayViewController: MBOverlayViewController, state: MBRecognizerResultState) {
        
        let recognizerRunnerViewController = overlayViewController.recognizerRunnerViewController as MBRecognizerRunnerViewController
        /** This is done on background thread */
        recognizerRunnerViewController.pauseScanning()
        
        
        
        var message: String = ""
        var title: String = ""
        
        if (self.barcodeRecognizer!.result.resultState == MBRecognizerResultState.valid) {
            title = "QR code"
            
            // Save the string representation of the code
            message = self.barcodeRecognizer!.result.stringData()
        }
        else if (self.pdf417Recognizer!.result.resultState == MBRecognizerResultState.valid) {
            title = "PDF417"
            
            // Save the string representation of the code
            message = self.pdf417Recognizer!.result.stringData()
        }
        
        /** Needs to be called on main thread beacuse everything prior is on background thread */
        DispatchQueue.main.async {
            // present the alert view with scanned results
            
            let alertController: UIAlertController = UIAlertController.init(title: title, message: message, preferredStyle: UIAlertControllerStyle.alert)
            
            let okAction: UIAlertAction = UIAlertAction.init(title: "OK", style: UIAlertActionStyle.default,
                                                             handler: { (action) -> Void in
                                                                self.dismiss(animated: true, completion: nil)
            })
            alertController.addAction(okAction)
            overlayViewController.present(alertController, animated: true, completion: nil)
        }
    }
    
    func overlayViewControllerDidTapClose(_ overlayViewController: MBOverlayViewController) {
        self.dismiss(animated: true, completion: nil)
    }
    
    func customOverlayViewControllerDidFinishScanning(_ customOverlay: CustomOverlay, state: MBRecognizerResultState) {
        self.overlayViewControllerDidFinishScanning(customOverlay, state: state)
    }
    
    func customOverlayViewControllerDidTapClose(customOverlay: CustomOverlay) {
        self.overlayViewControllerDidTapClose(customOverlay)
    }
    
    func barcodeOverlayViewControllerDidFinishScanning(_ barcodeOverlayViewController: MBBarcodeOverlayViewController, state: MBRecognizerResultState) {
        self.overlayViewControllerDidFinishScanning(barcodeOverlayViewController, state: state)
    }
    
    func barcodeOverlayViewControllerDidTapClose(_ barcodeOverlayViewController: MBBarcodeOverlayViewController) {
        self.overlayViewControllerDidTapClose(barcodeOverlayViewController)
    }
    
}
